gateway-api/src/gateway_api/time_duration.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+/// Converting from std::time::Duration to gateway_api::Duration is allowed,
+/// but we need to make sure that the incoming duration is valid according to
+/// GEP-2257.
+impl TryFrom<stdDuration> for Duration {
+    type Error = String;
+
+    fn try_from(d: stdDuration) -> Result<Self, Self::Error> {
+        // Check validity, and propagate any error if it's not.
+        is_valid(d)?;
+
+        // It's valid, so we can safely convert it to a gateway_api::Duration.
+        Ok(Duration(d))
+    }
+}
+
+impl Duration {
+    /// gateway_api::Duration::new creates a new gateway_api::Duration from
+    /// seconds and nanoseconds, but requires that the resulting Duration be
+    /// valid according to GEP-2257.
+    pub fn new(secs: u64, nanos: u32) -> Result<Self, String> {
+        let stddur = stdDuration::new(secs, nanos);
+
+        // Propagate errors if not valid, or unwrap the new Duration if all's
+        // well.
+        is_valid(stddur)?;
+        Ok(Self(stddur))
+    }
+
+    /// gateway_api::Duration::from_secs creates a new gateway_api::Duration
+    /// from seconds, but requires that the resulting Duration be valid
+    /// according to GEP-2257.
+    pub fn from_secs(secs: u64) -> Result<Self, String> {
+        Self::new(secs, 0)
+    }
+
+    /// gateway_api::Duration::from_micros creates a new gateway_api::Duration
+    /// from microseconds, but requires that the resulting Duration be valid
+    /// according to GEP-2257.
+    pub fn from_micros(micros: u64) -> Result<Self, String> {
+        let sec = micros / 1_000_000;
+        let ns = ((micros % 1_000_000) * 1_000) as u32;
+
+        Self::new(sec, ns)
+    }
+
+    /// gateway_api::Duration::from_millis creates a new gateway_api::Duration
+    /// from milliseconds, but requires that the resulting Duration be valid
+    /// according to GEP-2257.
+    pub fn from_millis(millis: u64) -> Result<Self, String> {
+        let sec = millis / 1_000;
+        let ns = ((millis % 1_000) * 1_000_000) as u32;
+
+        Self::new(sec, ns)
+    }
+
+    delegate! {
+        to self.0 {
+            /// gateway_api::Duration::as_secs returns the number of whole
+            /// seconds in the duration.
+            pub fn as_secs(&self) -> u64;
+
+            /// gateway_api::Duration::as_millis returns the number of whole
+            /// milliseconds in the duration.
+            pub fn as_millis(&self) -> u128;
+
+            /// gateway_api::Duration::as_micros returns the number of whole
+            /// microseconds in the duration.
+            pub fn as_nanos(&self) -> u128;
+
+            /// gateway_api::Duration::subsec_nanos returns the number of
+            /// nanoseconds in the duration that are not part of the whole
+            /// seconds.
+            pub fn subsec_nanos(&self) -> u32;
+
+            /// gateway_api::Duration::is_zero returns true if the duration is
+            /// zero.
+            pub fn is_zero(&self) -> bool;
+        }
+    }
```

Comment:
```suggestion

    /// The number of whole seconds in the duration
    pub fn as_secs(&self) -> u64 {
        self.0.as_secs()
    }

    /// The number of whole milliseconds in the duration
    pub fn as_millis(&self) -> u128 {
        self.0.as_millis()
    }

    /// The number of whole microseconds in the duration
    pub fn as_nanos(&self) -> u128 {
        self.0.as_nanos()
    }

    /// The number of nanoseconds in the duration that are
    /// not part of the whole seconds
    pub fn subsec_nanos(&self) -> u32 {
        self.0.subsec_nanos()
    }

    /// Indicates whether or not the duration is zero
    pub fn is_zero(&self) -> bool {
        self.0.is_zero()
    }
```

I 100% agree that it's _super_ annoying to have to type out delegations like that, but for the sake of *not* introducing a potentially unneeded dependency, I will absolutely volunteer to do it 😅

> **NOTE:** if this suggestion is accepted, the `use delegate::delegate;` will also need to be removed from line 1 and `delegate = "0.12.0"` from line 21 of `Cargo.toml`.

Reply:
I'm honestly torn on this one, so I'm going to collect some more opinions. 🙂

Reply:
Taking this one.